### PR TITLE
Tab completion

### PR DESCRIPTION
**Pull request by denCalderon**
This feature uses argcomplete package, its documentation can be found at https://kislyuk.github.io/argcomplete/

In order to enable it, please follow the next steps:
-Install: pip install argcomplete
-Register the script explicitly adding in e.g. .bashrc: eval "$(register-python-argcomplete hekit.py)"

### CODE REVIEW

File: kit/hekit.py [hunk 1]
```diff
@@ -9,10 +9,31 @@
...
+def parse_components(prefix, parsed_args, **kwargs):
+    parsed_args.config = load_config(parsed_args.config)
+    return get_components(parsed_args.config.repo_location)
+
+
+def parse_instances(prefix, parsed_args, **kwargs):
+    parsed_args.config = load_config(parsed_args.config)
+    comp_name_path = f"{parsed_args.config.repo_location}/{parsed_args.component}"
+    return get_instances(comp_name_path)
+
```
Comment (hamishun) [inline kit/hekit.py:35]:
```suggestion
def components_completer(prefix, parsed_args, **kwargs):
    parsed_args.config = load_config(parsed_args.config)
    return get_components(parsed_args.config.repo_location)


def instances_completer(prefix, parsed_args, **kwargs):
    parsed_args.config = load_config(parsed_args.config)
    comp_name_path = f"{parsed_args.config.repo_location}/{parsed_args.component}"
    return get_instances(comp_name_path)

```

File: kit/hekit.py [hunk 2]
```diff
@@ -9,10 +9,31 @@
...
+    try:
+        from argcomplete import autocomplete
+
+        autocomplete(parser)
+    except ImportError:
+        # ignore error if not installed
+        pass
+
+
+def parse_components(prefix, parsed_args, **kwargs):
```
Comment (hamishun) [inline kit/hekit.py:26]: These new functions could be placed into their own module and included if the autocomplete is successful in the try block

File: kit/hekit.py [hunk 3]
```diff
@@ -64,10 +85,16 @@ def parse_cmdline():
...
         "remove", description="removes/uninstalls components"
     )
-    parser_remove.add_argument("component", type=str, help="component to be removed")
-    parser_remove.add_argument("instance", type=str, help="instance to be removed")
+    parser_remove.add_argument(
+        "component", type=str, help="component to be removed"
+    ).completer = parse_components
+    parser_remove.add_argument(
+        "instance", type=str, help="instance to be removed"
+    ).completer = parse_instances
```
Comment (hamishun) [inline kit/hekit.py:93]: Will this not complain if the argcomplete fails to import?

**Review by hamishun — CHANGES_REQUESTED** [review]
Where is the documentation in the README.md?

File: kit/hekit.py (hunk 2 above)
Comment (denCalderon) [reply kit/hekit.py]: They were moved to tab_completion.py

File: kit/hekit.py (hunk 1 above)
Comment (denCalderon) [reply kit/hekit.py]: updated as requested

File: kit/hekit.py (hunk 3 above)
Comment (denCalderon) [reply kit/hekit.py]: I tested on the duckers build were argcomplete is not installed, and the python program can be executed correctly,  but it requires than the functions  (parse_components and parse_instances) were declared

File: kit/tab_completion.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
+#! /usr/bin/python3
+
```
Comment (hamishun) [inline kit/tab_completion.py:2]: This line is probably not required

**Review by hamishun — CHANGES_REQUESTED** [review]

File: kit/tab_completion.py (hunk 4 above)
Comment (denCalderon) [reply kit/tab_completion.py]: It was deleted

File: kit/tab_completion.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+from config import load_config
+from command_list import get_components, get_instances
+
+
+def enable_tab_completion(parser):
+    try:
+        from argcomplete import autocomplete
+
+        autocomplete(parser)
+    except ImportError:
```
Comment (jlhcrawford) [inline kit/tab_completion.py:13]: Maybe add a comment here or where this function is called saying argcomplete is optional. It's unusual to catch an exception and then pass.

**Review by jlhcrawford — CHANGES_REQUESTED** [review]
This is a general comment on the python code. I know most of the functions have self-documenting names but it's good practice to still have a comment documenting what the function does. This can be a few words or a sentence in a docstring at the top of the function.

File: kit/tab_completion.py (hunk 5 above)
Comment (denCalderon) [reply kit/tab_completion.py:17]: new comments were addes

File: kit/command_init.py [hunk 6]
```diff
@@ -68,19 +76,39 @@ def append_to_rc(path: str, content: str) -> None:
...
+    # New lines that will be added in the rc_file:
+    # 1-Add hekit directory as part of environmental variable PATH
+    content = f"PATH={args.hekit_root_dir}:$PATH"
+
+    # 2-Register hekit link and hekit.py script to enable tab completion
+    if not util.find_spec("argcomplete") == None:
+        register_py = f'eval "$(register-python-argcomplete hekit.py)"'
+        register_link = f'eval "$(register-python-argcomplete hekit)"'
+        content = "\n".join([content, register_py, register_link])
+
```
Comment (hamishun) [inline kit/command_init.py:112]: The test of argcomplete should not be in python it should be in POSIX shell if statement.
That way when a user decided afterwards that they suddenly do want tab completion they do not require to rerun hekit init.
```suggestion
    # 2-Register hekit link and hekit.py script to enable tab completion
    if not util.find_spec("argcomplete") == None:
        register_py = f'eval "$(register-python-argcomplete hekit.py)"'
        register_link = f'eval "$(register-python-argcomplete hekit)"'
        content = "\n".join([content, register_py, register_link])

```

**Review by hamishun — CHANGES_REQUESTED** [review]

File: docker/Dockerfile.base [hunk 7]
```diff
@@ -68,7 +68,8 @@ RUN wget https://github.com/Kitware/CMake/releases/download/v3.19.2/cmake-3.19.2
     rm -rf cmake_files
     # Install GMP, pip and python dependencies \
 RUN apt install -y libgmp-dev python3-pip && \
-    pip install toml
+    pip install toml && \
+    pip install argcomplete
```
[the remark this answers is not in the document]
Comment (jlhcrawford) [reply docker/Dockerfile.base:72]: We can use a requirements file and specify minimum versions since the number of python dependencies has doubled.

**Review by hamishun — APPROVED** [review]

**Pull request merged**